collector/uptrack.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"flag"
+	"io/ioutil"
+	"path"
+	"syscall"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	uptrackUpgradePlanPath = flag.String(
+		"collector.uptrack.upgrade_plan_path",
+		uptrackDefaultUpgradePlanPath(),
+		"Path where Uptrack stores its upgrade_plan YAML file.")
+)
+
+type uptrackCollector struct {
+	timeDesc      *prometheus.Desc
+	planCountDesc *prometheus.Desc
+}
+
+func init() {
+	Factories["uptrack"] = NewUptrackCollector
+}
+
+// The location of the upgrade_plan file is based on the currently running kernel.
+func uptrackDefaultUpgradePlanPath() string {
+	var uname syscall.Utsname
+	syscall.Uname(&uname)
+	return path.Join(
+		"/var/cache/uptrack",
```

Comment:
So what about this now? Given we decided not to scope the node-exporter to kernel-only data, this isn't out of scope in general. But I still prefer not to merge this because:

- It's not very commonly used
- AFAIK the ksplice+uptrack combination isn't FOSS
- This can be implemented via the textfile collector reasonably and simple

If anyone disagrees and still think this should get merged, please let me know or I'll close this in a week or so. Hope that's fine.

@EdSchouten I'm truly sorry we didn't communicate this earlier and you wasted your time on this :(